Source/Engine/Scripting/Coroutines/CoroutineSequence.h

```diff
@@ -0,0 +1,165 @@
+
+    /// <summary> Executes the code. </summary>
+    API_FUNCTION()
+    ScriptingObjectReference<CoroutineSequence> ThenRun(ScriptingObjectReference<CoroutineRunnable> runnable);
+
+    /// <summary>
+    /// Suspends the coroutine for the given number of seconds.
+    /// </summary>
+    /// <param name="seconds">
+    /// Minimal number of seconds to wait.
+    /// 0 is a valid value where the coroutine will be resumed on the closest time accumulation point.
+    /// </param>
+    API_FUNCTION()
+    ScriptingObjectReference<CoroutineSequence> ThenWaitSeconds(float seconds);
+
+    /// <summary> Suspends the coroutine for the given number of frames. </summary>
+    /// <param name="frames">
+    /// Minimal number of <b>full</b> frames to wait.
+    /// 0 is a valid value where the coroutine will be resumed on the closest frame accumulation point.
+    /// </param>
+    API_FUNCTION()
+    ScriptingObjectReference<CoroutineSequence> ThenWaitFrames(int32 frames);
+
+    /// <summary> Suspends the coroutine until the given suspend point. </summary>
+    ScriptingObjectReference<CoroutineSequence> ThenWaitForPoint(CoroutineSuspendPoint point);
+
+    /// <summary> Suspends the coroutine until the predicate is true. </summary>
+    API_FUNCTION()
+    ScriptingObjectReference<CoroutineSequence> ThenWaitUntil(ScriptingObjectReference<CoroutinePredicate> predicate);
+
+
+    /// <summary> QOL wrapper for running runnable action in native scripts. </summary>
+    ScriptingObjectReference<CoroutineSequence> ThenRunFunc(const Function<void()>& runnable);
+
+    /// <summary> QOL wrapper for waiting until predicate in native scripts. </summary>
+    ScriptingObjectReference<CoroutineSequence> ThenWaitUntilFunc(const Function<void(bool&)>& predicate);
+
+
+    using RunnableReference  = ScriptingObjectReference<CoroutineRunnable>;
+    using PredicateReference = ScriptingObjectReference<CoroutinePredicate>;
+
+    enum class StepType
+    {
+        None,
+        Run,
+        WaitSuspensionPoint,
+        WaitSeconds,
+        WaitFrames,
+        WaitUntil,
+    };
+
+    class Step final
+    {
+        // No std::variant in Cpp 14 :( Need to implement it manually.
+
+        union 
+        {
+            RunnableReference _runnable;
+            PredicateReference _predicate;
+
+            int32 _framesDelay;
+            float _secondsDelay;
+            CoroutineSuspendPoint _suspensionPoint;
+        };
+
+        StepType _type;
+
+    public:
+        explicit Step(RunnableReference&& runnable);
+        explicit Step(CoroutineSuspendPoint suspensionPoint);
+        explicit Step(int32 framesDelay);
+        explicit Step(float secondsDelay);
+        explicit Step(PredicateReference&& predicate);
+
+        Step(const Step& other);
+        Step(Step&& other) noexcept;
+
+    private:
+        void Clear();
+        void EmplaceCopy(const Step& other);
+        void EmplaceMove(Step&& other) noexcept;
```

Comment:
I think memory management for such structure is overenginnering. It can be data-only (maybe with frames delta, seconds delay and suspension point as union). Only `CoroutineSequence` and `CoroutineExecutor` use it.